### PR TITLE
Hybrid query optimize sortby vector

### PR DESCRIPTION
**Pull request by alonre24**
Avoid deep copying of the RSIndexResult subtree tree in hybrid vector reader when there is no need to compute the scores. This happens when:
"WITHSCORES" arg was not given, and there is a SORTBY clause.
This PR introduces a small change in the coordinator protocol which always sends the "WITHSCORES" arg - now it is send only if it is required - to apply this optimisation in cluster mode as well.

### CODE REVIEW

File: src/search_options.h [hunk 1]
```diff
@@ -72,7 +72,8 @@ typedef enum {
   Search_Verbatim = 0x02,
   Search_NoStopwrods = 0x04,
   Search_InOrder = 0x20,
-  Search_HasSlop = 0x200
+  Search_HasSlop = 0x200,
+  Search_IgnoreScores = 0x400
 } RSSearchFlags;
```
Comment (GuyAv46) [inline src/search_options.h:77]: Do we know why this enum skips bits?

**Review by GuyAv46 — DISMISSED** [review]
Very nice!

File: src/search_options.h (hunk 1 above)
Comment (alonre24) [reply src/search_options.h:77]: @ashtul do you know?

File: src/query.c [hunk 2]
```diff
@@ -649,7 +649,9 @@ static IndexIterator *Query_EvalVectorNode(QueryEvalCtx *q, QueryNode *qn) {
       return NULL;
     }
   }
-  IndexIterator *it = NewVectorIterator(q->sctx, qn->vn.vq, child_it, q->status);
+  // Check if we can ignore the document score
+  bool ignoreScores = q->opts->flags & Search_IgnoreScores;
+  IndexIterator *it = NewVectorIterator(q->sctx, qn->vn.vq, child_it, ignoreScores, q->status);
```
Comment (DvirDukhan) [inline src/query.c:654]: can we send `q` instead of extracting the `sctx`, `status`, `ignoreScores`?
`ignoreScores` is used to create the hybrid iterator only, so we can extract everything form `q`

File: src/hybrid_reader.h [hunk 3]
```diff
@@ -31,13 +31,14 @@ typedef struct {
...
   size_t numIterations;
+  bool ignoreScores;               // Ignore the document scores, only vector score matters.
 } HybridIterator;
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-IndexIterator *NewHybridVectorIterator(VecSimIndex *index, char *score_field, KNNVectorQuery query, VecSimQueryParams qParams, IndexIterator *child_it);
+IndexIterator *NewHybridVectorIterator(VecSimIndex *index, char *score_field, KNNVectorQuery query, VecSimQueryParams qParams, IndexIterator *child_it, bool ignoreScores);
```
Comment (DvirDukhan) [inline src/hybrid_reader.h:41]: once https://github.com/RediSearch/RediSearch/pull/2647 will be merged, this constructor signature will be unreadable. Please create a struct with everything we need for initializing the hybrid iterator

**Review by DvirDukhan — COMMENTED** [review]
well done
see the comments

File: src/query_parser/tokenizer.h [hunk 4]
```diff
@@ -2,8 +2,9 @@
 #define __QUERY_TOKENIZER_H__
 
 #include <stdlib.h>
+#include <stdbool.h>
 #include "../tokenize.h"
-#include "vector_index.h"
+#include "VecSim/vec_sim_common.h"
```
Comment (DvirDukhan) [inline src/query_parser/tokenizer.h:7]: why the change here?

**Review by DvirDukhan — COMMENTED** [review]
looks good

one question

File: src/query_parser/tokenizer.h (hunk 4 above)
Comment (alonre24) [reply src/query_parser/tokenizer.h:7]: There was cyclic include:
tokenizer.h -> vector_index.h -> query_ctx.h -> query_param.h -> tokenizer.h
Actually now that I look, we shouldn't include query_param.h from query_ctx.h. 
But anyway, I think that it is safer that tokenizer.h will include vec_sim_common.h (to avoid cyclic includes in the future), no?

Comment (GuyAv46) [reply src/query_parser/tokenizer.h:7]: I included `vector_index.h` for `VecSimRawParam` but since we moved this struct to the library we can change this. On the other hand maybe we want to include files from the library through `vector_index.h` anyway

**Review by DvirDukhan — DISMISSED** [review]

File: src/query.c [hunk 5]
```diff
@@ -649,7 +649,9 @@ static IndexIterator *Query_EvalVectorNode(QueryEvalCtx *q, QueryNode *qn) {
       return NULL;
     }
   }
-  IndexIterator *it = NewVectorIterator(q->sctx, qn->vn.vq, child_it, q->status);
+  // Check if we can ignore the document score
+  bool ignoreScores = q->opts->flags & Search_IgnoreScores;
+  IndexIterator *it = NewVectorIterator(q, qn->vn.vq, child_it);
```
Comment (GuyAv46) [inline src/query.c:654]: seems like you don't use this variable

Comment (alonre24) [reply src/query.c:654]: right, thanks!

File: src/aggregate/aggregate_request.c [hunk 6]
```diff
@@ -187,6 +187,10 @@ static int handleCommonArgs(AREQ *req, ArgsCursor *ac, QueryError *status, int a
     if ((parseSortby(arng, ac, status, req->reqflags & QEXEC_F_IS_SEARCH)) != REDISMODULE_OK) {
       return ARG_ERROR;
     }
+    // If we have sort by clause and don't need to return scores, we can avoid computing them.
+    if (req->reqflags & QEXEC_F_IS_SEARCH && !(QEXEC_F_SEND_SCORES & req->reqflags)) {
```
Comment (ashtul) [inline src/aggregate/aggregate_request.c:191]: why QEXEC_F_IS_SEARCH?

**Review by ashtul — COMMENTED** [review]
a few comments

File: coord/src/module.c [hunk 7]
```diff
@@ -1646,9 +1652,10 @@ int FlatSearchCommandHandler(RedisModuleBlockedClient *bc, RedisModuleString **a
     MRCommand_ReplaceArg(&cmd, 0, "_FT.PROFILE", sizeof("_FT.PROFILE") - 1);
   }
 
-  // adding the WITHSCORES option anyway immediately after the query.
-  // Worst case it will appears twice.
-  MRCommand_AppendArgsAtPos(&cmd, 3 + req->profileArgs, 1, "WITHSCORES");
+  // adding the WITHSCORES option only if there is no SORTBY (hence the score is the default sort key)
+  if (!req->withSortby) {
+    MRCommand_AppendArgsAtPos(&cmd, 3 + req->profileArgs, 1, "WITHSCORES");
```
Comment (MeirShpilraien) [inline coord/src/module.c:1657]: How do you know `WITHSCORES` is not requested by the user?

Comment (MeirShpilraien) [reply coord/src/module.c:1657]: Ohh if it requested it will be passed as is :+1:

File: src/vector_index.c [hunk 8]
```diff
@@ -100,17 +101,24 @@ IndexIterator *NewVectorIterator(RedisSearchCtx *ctx, VectorQuery *vq, IndexIter
...
       }
-      return NewHybridVectorIterator(vecsim, vq->scoreField, vq->knn, qParams, child_it);
+      HybridIteratorParams hParams = {.index = vecsim,
+                                      .query = vq->knn,
+                                      .qParams = qParams,
+                                      .vectorScoreField = vq->scoreField,
+                                      .ignoreDocScore = q->opts->flags & Search_IgnoreScores,
+                                      .childIt = child_it
+      };
+      return NewHybridVectorIterator(hParams);
```
Comment (MeirShpilraien) [inline src/vector_index.c:121]: nitpick, I do not see why it is better to wrap it with a struct, you could have pass all those argument to the function and use new lines and indentations to make it look almost the same as it look now. It could be useful if you would have initialised it somewhere else and then use it here...

**Review by MeirShpilraien — APPROVED** [review]
:+1: one comment, not a blocker.

File: src/vector_index.c (hunk 8 above)
Comment (alonre24) [reply src/vector_index.c:121]: We decided to do it because we added several times new parameters to this function and had to change the signature every time (and now we can just extend the struct). But overall I agree that it doesn't really matter...
